### PR TITLE
Rdrp 979 tree to list

### PR DESCRIPTION
**Pull request by zorge69**
## Converting a directory tree specified as a Python dict into a list of full paths, so they can be created one by one.

*Also, convert_formtype was repeated twice - removed one copy of it.*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/utils/helpers.py
Comment (AnneONS) [inline src/utils/helpers.py:114]: Have you removed this cos it's not used and we have another function doing the same thing in construction?

File: tests/test_utils/test_helpers.py
Comment (AnneONS) [inline tests/test_utils/test_helpers.py:7]: haven't you removed the convert formtype function?

**Review by AnneONS — COMMENTED** [review]
Looks good so far! But I'm interested to know how you'd actually use this to create directories.

File: tests/test_utils/test_helpers.py
Comment (zorge69) [reply tests/test_utils/test_helpers.py:7]: No, it's still there. I just had to wrap a long line.

File: tests/test_utils/test_helpers.py
Comment (cbrowning04) [inline tests/test_utils/test_helpers.py]: A bit tedious but could you add docstrings here.

File: tests/test_utils/test_helpers.py
Comment (cbrowning04) [inline tests/test_utils/test_helpers.py]: Can this be split into multiple tests? E.g., 

1. General passing test
2. A failing test (where it raises)

May I also suggest adding a test that covers all of the parameters? For example, separator does not get tested with
anything but the default.

File: tests/test_utils/test_helpers.py
Comment (cbrowning04) [inline tests/test_utils/test_helpers.py]: Please add a matching message for the error raise here.

**Review by cbrowning04 — CHANGES_REQUESTED** [review]
Hi @zorge69 . Good stuff, just a few comments on the tests here.

**Comment by zorge69** [comment]
> Looks good so far! But I'm interested to know how you'd actually use this to create directories.
Added the directories as a dict and a main file to create them: https://github.com/ONSdigital/research-and-development/blob/RDRP-979_make_directories/make_dirs_main.py

File: src/utils/helpers.py
Comment (zorge69) [reply src/utils/helpers.py:114]: I removed it, because convert_formtype function was repeated twice, two identical blocks of code, starting from https://github.com/ONSdigital/research-and-development/blob/3c583abd760800b1a30b8c7d2fab418d3057e924/src/utils/helpers.py#L69 and from https://github.com/ONSdigital/research-and-development/blob/3c583abd760800b1a30b8c7d2fab418d3057e924/src/utils/helpers.py#L92

**Comment by zorge69** [comment]
> Looks good so far! But I'm interested to know how you'd actually use this to create directories.

https://github.com/ONSdigital/research-and-development/blob/d266a871a82a77d82d0e442118ea212f3dda546c/make_dirs_main.py#L1

File: tests/test_utils/test_helpers.py
Comment (zorge69) [reply tests/test_utils/test_helpers.py:7]: I removed a duplicated block of code.

**Review by AnneONS — COMMENTED** [review]
Hi George, it looks like you've got the changes from other branches in here as well- so presumably we can't merge in this one until we've merged in the other branches (as doing so will put in those changes too.

File: tests/test_utils/test_helpers.py
Comment (zorge69) [reply tests/test_utils/test_helpers.py]: Updated doc strings, added type hints

**Comment by zorge69** [comment]
> Hi George, it looks like you've got the changes from other branches in here as well- so presumably we can't merge in this one until we've merged in the other branches (as doing so will put in those changes too.

Yes, unfortunately. Let's wait for s3 mods to be merged.

File: tests/test_utils/test_helpers.py
Comment (zorge69) [reply tests/test_utils/test_helpers.py]: Added a negative test that checks that the error message matches.
Separator: we will never change it, so I hardcoded it.

File: tests/test_utils/test_helpers.py
Comment (zorge69) [reply tests/test_utils/test_helpers.py]: done

**Review by cbrowning04 — APPROVED** [review]

**Pull request merged**
